services/eth_gateway.go

```diff
@@ -0,0 +1,67 @@
+package services
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// TODO: Find etherum node to connect to.
+const ethUrl = "ws://0.0.0.0"
+
+// Singleton client
+var client *ethclient.Client
+var mtx sync.Mutex
+
+func sharedClient() (c *ethclient.Client, err error) {
+	if client != nil {
+		return client, nil
+	}
+
+	// check-lock-check pattern to avoid excessive locking.
+	mtx.Lock()
+	defer mtx.Unlock()
+
+	if client != nil {
+		c, err = ethclient.Dial(ethUrl)
```

Comment:
ethclient is going to be nil in this case.

Reply:
ethclient is the package name (imported on line 11). Unfortunate naming...